File: backend/selko/api/routes/sender_rules.py
"""Sender rules API routes."""

import logging
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from supabase import Client

from selko.api.deps import get_authenticated_client
from selko.api.schemas.events import SenderRuleRequest, SenderRuleResponse

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/sender-rules", tags=["sender-rules"])


@router.get("", response_model=list[SenderRuleResponse])
async def list_sender_rules(
    client: Annotated[Client, Depends(get_authenticated_client)],
) -> list[SenderRuleResponse]:
    """List user's sender rules."""
    try:
        user_id = client.auth.get_user().user.id
        
        result = client.table("sender_rules").select("*").eq(
            "user_id", user_id
        ).order("created_at").execute()
        
        return [
            SenderRuleResponse(
                id=rule["id"],
                sender_domain=rule.get("sender_domain"),
                sender_email=rule.get("sender_email"),
                action=rule["action"],
                created_at=rule["created_at"],
            )
            for rule in result.data
        ]
        
    except Exception as e:
        logger.error(f"Failed to list sender rules: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("", response_model=SenderRuleResponse)
async def create_sender_rule(
    request: SenderRuleRequest,
    client: Annotated[Client, Depends(get_authenticated_client)],
) -> SenderRuleResponse:
    """Create rule (auto_approve or ignore)."""
    try:
        user_id = client.auth.get_user().user.id
        
        # Validate at least one of domain or email is set
        if not request.sender_domain and not request.sender_email:
            raise HTTPException(
                status_code=400,
                detail="Either sender_domain or sender_email must be provided"
            )
        
        result = client.table("sender_rules").insert({
            "user_id": user_id,
            "sender_domain": request.sender_domain,
            "sender_email": request.sender_email,
            "action": request.action,
        }).execute()
        
        rule = result.data[0]
        
        return SenderRuleResponse(
            id=rule["id"],
            sender_domain=rule.get("sender_domain"),
            sender_email=rule.get("sender_email"),
            action=rule["action"],
            created_at=rule["created_at"],
        )
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to create sender rule: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/{rule_id}")
async def delete_sender_rule(
    rule_id: UUID,
    client: Annotated[Client, Depends(get_authenticated_client)],
):
    """Delete rule."""
    try:
        user_id = client.auth.get_user().user.id
        
        # Verify ownership
        rule_result = client.table("sender_rules").select("user_id").eq(
            "id", str(rule_id)
        ).single().execute()
        
        if rule_result.data["user_id"] != user_id:
            raise HTTPException(status_code=403, detail="Not authorized")
        
        client.table("sender_rules").delete().eq("id", str(rule_id)).execute()
        
        return {"status": "deleted"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to delete sender rule: {e}")
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/lib/services/sender-rules.js
import { supabase } from '$lib/supabase.js';
import { parseSupabaseError } from '$lib/errors.js';

/**
 * @typedef {'auto_approve' | 'ignore'} SenderRuleAction
 */

/**
 * @typedef {Object} SenderRule
 * @property {string} id - UUID
 * @property {string} user_id - UUID
 * @property {string | null} sender_domain - Domain to match (e.g., "example.com")
 * @property {string | null} sender_email - Email to match (takes precedence over domain)
 * @property {SenderRuleAction} action
 * @property {string} created_at
 * @property {string} updated_at
 */

/**
 * Fetch all sender rules for the current user
 * @returns {Promise<{data: SenderRule[], error: import('$lib/errors.js').SupabaseError | null}>}
 */
export async function fetchSenderRules() {
	try {
		const { data, error } = await supabase
			.from('sender_rules')
			.select('*')
			.order('created_at', { ascending: false });

		if (error) throw error;

		return { data: data ?? [], error: null };
	} catch (error) {
		return { data: [], error: parseSupabaseError(error) };
	}
}

/**
 * Create a new sender rule
 * @param {Object} rule
 * @param {string} [rule.sender_domain] - Domain to match (e.g., "example.com")
 * @param {string} [rule.sender_email] - Email to match (takes precedence over domain)
 * @param {SenderRuleAction} rule.action - Action to take
 * @returns {Promise<{data: SenderRule | null, error: import('$lib/errors.js').SupabaseError | null}>}
 */
export async function createSenderRule(rule) {
	try {
		// Validate that at least one of domain or email is provided
		if (!rule.sender_domain && !rule.sender_email) {
			throw new Error('Either sender_domain or sender_email must be provided');
		}

		const {
			data: { user }
		} = await supabase.auth.getUser();
		if (!user) throw new Error('Not authenticated');

		const { data, error } = await supabase
			.from('sender_rules')
			.insert({ ...rule, user_id: user.id })
			.select()
			.single();

		if (error) throw error;

		return { data, error: null };
	} catch (error) {
		return { data: null, error: parseSupabaseError(error) };
	}
}

/**
 * Ignore a sender retroactively: creates the ignore rule and, in the same
 * atomic server-side call, rejects their pending events in the New lane AND
 * discards their proposals in the Changes lane.
 * @param {string} senderEmail - The sender's email address
 * @returns {Promise<{data: {rejected_new: number, discarded_changes: number} | null, error: import('$lib/errors.js').SupabaseError | null}>}
 */
export async function ignoreSenderRetroactive(senderEmail) {
	try {
		const { data, error } = await supabase.rpc('ignore_sender_and_reject_pending', {
			p_sender_email: senderEmail
		});
		if (error) throw error;
		return { data, error: null };
	} catch (error) {
		return { data: null, error: parseSupabaseError(error) };
	}
}

/**
 * Delete a sender rule
 * @param {string} ruleId - The rule UUID
 * @returns {Promise<{error: import('$lib/errors.js').SupabaseError | null}>}
 */
export async function deleteSenderRule(ruleId) {
	try {
		const { error } = await supabase.from('sender_rules').delete().eq('id', ruleId);

		if (error) throw error;

		return { error: null };
	} catch (error) {
		return { error: parseSupabaseError(error) };
	}
}

/**
 * Check if a sender matches any rule
 * @param {string} senderEmail - The sender's email address
 * @returns {Promise<{data: SenderRule | null, error: import('$lib/errors.js').SupabaseError | null}>}
 */
export async function findMatchingRule(senderEmail) {
	try {
		const domain = senderEmail.split('@')[1];

		// Check for exact email match first, then domain match
		const { data, error } = await supabase
			.from('sender_rules')
			.select('*')
			.or(`sender_email.eq.${senderEmail},sender_domain.eq.${domain}`)
			.order('sender_email', { ascending: false, nullsFirst: false }) // Email matches first
			.limit(1)
			.maybeSingle();

		if (error) throw error;

		return { data, error: null };
	} catch (error) {
		return { data: null, error: parseSupabaseError(error) };
	}
}
